src/raftstore/store/worker/apply.rs

```diff
@@ -885,26 +952,29 @@ impl ExecContext {
 impl ApplyDelegate {
     // Only errors that will also occur on all other stores should be returned.
     fn exec_raft_cmd(
-        &mut self,
-        ctx: &mut ApplyContext,
-    ) -> Result<(RaftCmdResponse, Option<ExecResult>)> {
-        let req = Rc::clone(&ctx.exec_ctx.as_ref().unwrap().req);
+        self,
+        ctx: ApplyContext,
+    ) -> Box<Future<Item = RaftCmdFutureItem, Error = FutureError> + Send> {
+        let mut req = ctx.exec_ctx.as_ref().unwrap().req.clone();
         // Include region for stale epoch after merge may cause key not in range.
         let include_region =
             req.get_header().get_region_epoch().get_version() >= self.last_merge_version;
-        check_region_epoch(&req, &self.region, include_region)?;
+        if let Err(e) = check_region_epoch(&req, &self.region, include_region) {
+            return box future::err((self, ctx, e));
+        }
+
         if req.has_admin_request() {
-            self.exec_admin_cmd(ctx, req.get_admin_request())
+            box self.exec_admin_cmd(ctx, req.take_admin_request())
         } else {
-            self.exec_write_cmd(ctx, req.get_requests())
+            box self.exec_write_cmd(ctx, req.take_requests().into_vec())
         }
     }
 
     fn exec_admin_cmd(
-        &mut self,
-        ctx: &mut ApplyContext,
-        request: &AdminRequest,
-    ) -> Result<(RaftCmdResponse, Option<ExecResult>)> {
+        self,
+        ctx: ApplyContext,
+        request: AdminRequest,
+    ) -> Box<Future<Item = RaftCmdFutureItem, Error = FutureError> + Send> {
```

Comment:
this box can be just removed to use impl future because the return type is always `Then<..>`